src/text-buffer.coffee

```diff
@@ -1199,19 +1200,33 @@ class TextBuffer
   ###
 
   # Public: Save the buffer.
-  save: ->
-    @saveAs(@getPath())
+  save: (options) ->
+    @saveAs(@getPath(), options)
 
   # Public: Save the buffer at a specific path.
   #
   # * `filePath` The path to save at.
-  saveAs: (filePath) ->
+  saveAs: (filePath, options) ->
     unless filePath then throw new Error("Can't save buffer with no file path")
 
     @emitter.emit 'will-save', {path: filePath}
     @emit 'will-be-saved', this if Grim.includeDeprecatedAPIs
     @setPath(filePath)
-    @file.writeSync(@getText())
+
+    if options?.backup and @file.existsSync()
+      backupFilePath = filePath + '~'
+      backupFilePath += '~' while fs.existsSync(backupFilePath)
+      fs.writeFileSync(backupFilePath, @file.readSync())
```

Comment:
It's worth using [fdatasync](http://linux.die.net/man/2/fdatasync) to make sure the backup file is on disk before truncating the main file.  Otherwise there's probably a couple-second window where, if you lose power, the user will find that the backup file isn't on disk yet and the main file's metadata is truncated.  Total loss.

Doing this after calling writeFileSync should be good enough:

```js
fd = fs.openSync(path, 'w')
fs.fdatasyncSync(fd) 
fs.closeSync(fd)
```

If you just created the backup file, you'll need to ensure its dirent is on disk too.  Otherwise, there's no point to ensuring the file contents are written!

```js
path = require 'path'
parent = path.dirname(@file.getPath())
fd = fs.openSync(parent, 'r')  // 'r' is important, 'w' doesn't work for dirs
fs.fdatasyncSync(fd) 
fs.closeSync(fd)
```

Just stumbled across this, matches my findings: http://www.daveeddy.com/2013/03/26/synchronous-file-io-in-nodejs/